content/academy/expert_scraping_with_apify/solutions/handling_migrations.md

```diff
@@ -0,0 +1,243 @@
+            data: {
+                ...data,
+                description: element.text().trim(),
+            },
+        },
+    });
+};
+
+exports.handleOffers = async ({ $, request }, dataset) => {
+    const { data } = request.userData;
+
+    const { asin } = data;
+
+    for (const offer of $('#aod-offer')) {
+        // For each offer, add 1 to the ASIN's
+        // offer count
+        tracker.addToASIN(asin);
+
+        const element = $(offer);
+
+        await dataset.pushData({
+            ...data,
+            sellerName: element.find('div[id*="soldBy"] a[aria-label]').text().trim(),
+            offer: element.find('.a-price .a-offscreen').text().trim(),
+        });
+    }
+};
+```
+
+## [](#persisting-state) Persisting state
+
+The **persistState** event is automatically fired (by default) every 60 seconds by the Apify SDK while the actor is running, and is also fired when the **migrating** event occurs.
+
+In order to persist our ASIN tracker object, let's use the `Apify.events.on` function to listen for the **persistState** event and store it in the key-value store each time it is emitted.
+
+```JavaScript
+// asinTracker.js
+const Apify = require('apify');
+// We've updated our constants.js file to include the name
+// of this new key in the key-value store
+const { ASIN_TRACKER } = require('./constants');
+
+class ASINTracker {
+    constructor() {
+        this.state = {};
+
+        Apify.events.on('persistState', async () => {
+            await Apify.setValue(ASIN_TRACKER, this.state);
+        });
+
+        setInterval(() => console.log(this.state), 10000);
+    }
+
+    addASIN(asin) {
+        this.state[asin] = 0;
+    }
+
+    addToASIN(asin) {
+        this.state[asin]++;
+    }
+}
+
+module.exports = new ASINTracker();
+```
+
+## [](#handling-resurrections) Handling resurrections
+
+Great! So now our state will be persisted every 60 seconds in the key-value store. However, we're not done. Let's say that the actor migrates and is resurrected. We never actually update the `state` variable of our `ASINTracker` class with the state stored in the key-value store, so as our code currently stands, we still don't support state-persistence on migrations.
+
+In order to fix this, let's create a method called `checkStore` which will be called at the very beginning of the actor's run, and will check the key-value store for a previous state under the key **ASIN-TRACKER**. If a previous state does live there, then it will update the class' `state` variable with the value read from the key-value store:
+
+```JavaScript
+// asinTracker.js
+const Apify = require('apify');
+const { ASIN_TRACKER } = require('./constants');
+
+class ASINTracker {
+    constructor() {
+        this.state = {};
+
+        Apify.events.on('persistState', async () => {
+            await Apify.setValue(ASIN_TRACKER, this.state);
+        });
+
+        setInterval(() => console.log(this.state), 10000);
+    }
+
+    async checkStore() {
```

Comment:
I would suggest more neutral name like `load()` or `initialize()`. The info that we are persisting in store is implementation detail.